File: 01-poo-ac/src/impl/thermostat-samsung.ts
import { CHANGE_TEMPERATURE_SERVICE, SEND_MESSAGE_SERVICE } from "data";
import { ChangeTemperature } from "services-interface/change-temperature";
import { SendMessage } from "services-interface/send-message";
import { Thermostat } from "interfaces/thermostat";
import { BehaviorSubject, Subscription } from "rxjs";
import { Inject, Service } from "typedi";
import { DestinyMessage } from "destiny-message";

@Service()
export class ThermostatSamsung implements Thermostat {
    changedTemperature: BehaviorSubject<number> = new BehaviorSubject<number>(20);
    changedTemperatureSubscription: Subscription;
    state: string = "off";

    constructor(
        @Inject(CHANGE_TEMPERATURE_SERVICE) private changeTemperatureService: ChangeTemperature,
        @Inject(SEND_MESSAGE_SERVICE) private sendMessageService: SendMessage
    ) {
        this.initService();
        this.sendMessageService.sendMessage({ destiny: DestinyMessage.ACTIONS, message: 'Thermostat Samsung intialized' } );
    }

    changeTemperature(temperature: number): void {
        throw new Error("Method not implemented.");
    }

    initService() {
        this.changedTemperatureSubscription =
            this.changeTemperatureService.changedTemperature.subscribe(this.whenTemperatureChanged); 
    }

    whenTemperatureChanged = (temperature: number): void => {
        this.state = 'on';
        this.changedTemperature.next(temperature);
    }

    finalize() {
        this.changedTemperatureSubscription.unsubscribe();
        this.sendMessageService.sendMessage({ destiny: DestinyMessage.ACTIONS, message: 'Thermostat Samsung finalized' } );
    }
}

File: Aire acondicionado/Implement/ThermostatSamsung.cs
using System;
using System.Reactive.Subjects;
using System.Reactive.Disposables;
using Aire_acondicionado.Constant;
using Aire_acondicionado.Interface;
using Aire_acondicionado.Services_interface;

namespace AireAcondicionado.Implement
{
    public class ThermostatSamsung : IThermostat
    {
        private readonly IChangeTemperature changeTemperatureService;
        private IDisposable changedTemperatureSubscription;
        public BehaviorSubject<double> ChangedTemperature => new BehaviorSubject<double>(20);
        private string state = "Thermostat OFF";

        public ThermostatSamsung(
            IChangeTemperature changeTemperatureService)
        {
            this.changeTemperatureService = changeTemperatureService;

            InitService();
        }

        private void InitService()
        {
            changedTemperatureSubscription = changeTemperatureService.ChangedTemperature.Subscribe(WhenTemperatureChanged);
        }

        private void WhenTemperatureChanged(double temperature)
        {
            state = "Thermostat ON";
            ChangedTemperature.OnNext(temperature);
        }

        public void ChangeTemperature(double temperature)
        {
            throw new NotImplementedException();
        }

        public void Finalize()
        {
            changedTemperatureSubscription.Dispose();
        }
    }
}
